relay-general/src/protocol/measurements.rs

```diff
@@ -0,0 +1,199 @@
+use regex::Regex;
+
+use crate::types::{Annotated, Error, FromValue, Object, Value};
+
+lazy_static::lazy_static! {
+    static ref MEASUREMENT_NAME: Regex = Regex::new("^[a-z0-9-._]+$").unwrap();
+}
+
+#[derive(Clone, Debug, Default, PartialEq, Empty, FromValue, ToValue, ProcessValue)]
+#[cfg_attr(feature = "jsonschema", derive(JsonSchema))]
+pub struct Measurement {
+    /// Value of observed measurement value
+    #[metastructure(required = "true")]
+    pub value: Annotated<f64>,
+}
+
+// TODO: need bag_size?
+#[derive(Clone, Debug, Default, PartialEq, Empty, ToValue, ProcessValue)]
+#[cfg_attr(feature = "jsonschema", derive(JsonSchema))]
+pub struct Measurements(pub Object<Measurement>);
+
+impl FromValue for Measurements {
+    fn from_value(value: Annotated<Value>) -> Annotated<Self> {
+        match value {
+            Annotated(Some(Value::Object(items)), mut measurement_meta) => {
+                let mut measurements = Object::<Measurement>::new();
+
+                for (raw_name, raw_observed_value) in items.into_iter() {
+                    let observed_value: Annotated<f64> = match raw_observed_value {
+                        Annotated(Some(Value::Object(bag)), mut object_meta) => {
+                            match bag.get("value") {
+                                Some(Annotated(Some(Value::I64(value)), meta)) => {
+                                    Annotated(Some(*value as f64), meta.clone())
+                                }
+                                Some(Annotated(Some(Value::U64(value)), meta)) => {
+                                    Annotated(Some(*value as f64), meta.clone())
+                                }
+                                Some(Annotated(Some(Value::F64(value)), meta)) => {
+                                    Annotated(Some(*value), meta.clone())
+                                }
+                                Some(Annotated(value, meta)) => {
+                                    measurement_meta.add_error(Error::expected(
+                                        format!(
+                                            "measurement value for '{}' to be a number",
+                                            raw_name
+                                        )
```

Comment:
I believe that this will be covered by the previous comment, but for completeness:

Those expectations should usually be static strings containing a fragment describing the desired format. Since measurements always have to be numbers, you can just put `Error::expected("a number")`. Annotation rendering will take care of the rest.